fix: Do not change default security groups during EKS control plane reconcile

dlipovetsky opened the pull request:
<!-- Thanks for this PR! If this is your first PR please read the [contributing guide](../CONTRIBUTING.md) -->
<!-- If this PR is still work-in-progress and is being open for visibility please prefix the title with `WIP:` -->

**What type of PR is this?**

<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
/kind regression
/kind support
-->
/kind bug

**What this PR does / why we need it**:
For EKS clusters, #3028 enabled the bastion security group only when the bastion is enabled. However, the implementation introduced a bug: every reconcile appends to a package-level variable!

<!-- Enter a description of the change and why this change is needed -->

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
I'd like to add a pair of unit tests, but haven't done that yet.

**Checklist**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR in which case these can be deleted -->

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests
- [ ] adds or updates e2e tests

pydctw commented:
The changes looks fine. As you mentioned, it will be great to have a unit test for this.

Also, I am interested to learn how you found the bug.

dlipovetsky commented:
> The changes looks fine. As you mentioned, it will be great to have a unit test for this.

I'll work on one.

> Also, I am interested to learn how you found the bug.

I wanted to add the "no bastion security group" feature to the EC2 cluster flavor (https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3482), and to do that I read through this code. That's when I spotted it.

Package-level variables can be tricky to use. There's even a linter (https://github.com/leighmcculloch/gochecknoglobals) for them.

A constant would be great, but we can't have one.

I could go a bit further in this PR: I could move the variable into a function, and replace every variable reference with a function call. I don't foresee a noticeable impact on performance.

sedefsavas commented:
LGTM as is.

dlipovetsky commented:
I've made some progress on creating a unit test for this, and I'd appreciate some more time to finish it.

The AWSManagedControlPlane controller did not have existing tests, so it's taken me some time to set up the scaffolding.

dlipovetsky commented:
I wanted to create a failing test for the bug without modifying the implementation, but the only way to do that is to complete two normal reconciles of the AWSManagedControlPlane. To test a normal reconcile, I would have to mock various AWS services, and that would require extensive changes to AWSManagedControlPlane.

If that's something we want anyway, I can contribute it in a separate PR, and then use it to create the failing test. But given the bug is relatively easy to understand, I think we can choose to skip a failing test.

dlipovetsky commented:
I'm sorry, this wasn't a clean update to the PR. I should have added another commit with the unit test, but instead I squashed it into the existing commit, _and_ I rebased on main :disappointed:.

dlipovetsky commented:
It would be good to get this fix in to release 1.5

/cherry-pick release-1.5

sedefsavas commented:
/lgtm

Skarlso commented on controlplane/eks/controllers/awsmanagedcontrolplane_controller.go:
```diff
@@ -57,11 +57,21 @@ const (
+
+// securityGroupRolesForControlPlane returns the security group roles determined by the control plane configuration.
+func securityGroupRolesForControlPlane(scope *scope.ManagedControlPlaneScope) []infrav1.SecurityGroupRole {
+	// Copy to ensure we do not modify the package-level variable.
+	roles := make([]infrav1.SecurityGroupRole, len(defaultEKSSecurityGroupRoles))
+	copy(roles, defaultEKSSecurityGroupRoles)
+
+	if scope.Bastion().Enabled {
+		roles = append(roles, infrav1.SecurityGroupBastion)
```
```suggestion
	// Copy to ensure we do not modify the package-level variable.
	roles := append([]infrav1.SecurityGroupRole{}, infrav1.SecurityGroupEKSNodeAdditional)
	if scope.Bastion().Enabled {
		roles = append(roles, infrav1.SecurityGroupBastion)
```

Does the same thing, without the copy need. :)

dlipovetsky replied:
I see your point, but I think we want to keep `defaultEKSSecurityGroupRoles` as a reminder of the default set of security group roles.

Skarlso replied:
Sure, if you really want that, you can do:

```go
roles := append([]infrav1.SecurityGroupRole{}, defaultEKSSecurityGroupRoles...)
```

Which will also keep the original. :)

dlipovetsky replied:
@pydctw and I also considered using `append` here, but we settled on using `copy`. I'll resolve this thread for now, but if you feel very strongly, please reopen.

Skarlso replied:
It's fine. :) I'm just pedantic. :D

dlipovetsky replied:
No problem at all :smile:  I am also pedantic! ... Just not on my own PRs :joy:

Skarlso commented on controlplane/eks/controllers/awsmanagedcontrolplane_controller_unit_test.go:
```diff
@@ -0,0 +1,63 @@
+/*
```
Nice test additions. :)

dlipovetsky replied:
Thanks for the review!

richardcase commented:
/milestone v1.5.0

richardcase commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardcase commented:
This change looks good, thanks @dlipovetsky 👍

The e2e tests are passing as well.

/approve

dlipovetsky commented:
I'll take a look at the test failure.

dlipovetsky commented:
The linter is complaining. It wasn't earlier, not sure what's changed.

dlipovetsky commented:
The test uses golangci-lint v1.47.2

```
make -C hack/tools bin/golangci-lint
make[1]: Entering directory '/home/prow/go/src/sigs.k8s.io/cluster-api-provider-aws/hack/tools'
mkdir -p bin
go build -tags=tools -o bin/golangci-lint github.com/golangci/golangci-lint/cmd/golangci-lint
go: downloading github.com/golangci/golangci-lint v1.47.2
```

But the branch uses v1.45.2:

```
> grep golangci-lint hack/tools/go.mod
	github.com/golangci/golangci-lint v1.45.2
```

I am surprised. The only way to merge this, or any PR that hits this issue, is to rebase on `main`. Frustratingly, that removes the lgtm label.

sedefsavas commented:
/lgtm

dlipovetsky commented:
Rebased on main to get the new linter version. Surprisingly, `make lint` passed locally, although I expected it to fail, because the test failed with this branch and the linter version from `main`. :thinking: :

dlipovetsky commented:
I don't know why the linter is complaining. I can't reproduce this locally. I'm also confused by the complaint about `pkg/cloud/services/s3/s3_test.go`, which I haven't touched in this PR.

```
level=info msg="[runner] linters took 1m1.207464569s with stages: goanalysis_metalinter: 1m1.1159032s"
pkg/internal/rate/rate.go:213: File is not `gofmt`-ed with `-s` (gofmt)
//   r := lim.ReserveN(time.Now(), 1)
//   if !r.OK() {
//     // Not allowed to act! Did you remember to set lim.burst to be > 0 ?
//     return
//   }
//   time.Sleep(r.Delay())
//   Act()
pkg/cloud/services/s3/mock_stsiface/doc.go:17: File is not `gofmt`-ed with `-s` (gofmt)
// Run go generate to regenerate this mock.
pkg/cloud/services/s3/mock_s3iface/doc.go:17: File is not `gofmt`-ed with `-s` (gofmt)
// Run go generate to regenerate this mock.
api/v1beta1/defaults.go:8: File is not `gofmt`-ed with `-s` (gofmt)
    http://www.apache.org/licenses/LICENSE-2.0
controlplane/eks/controllers/helpers_test.go:8: File is not `gofmt`-ed with `-s` (gofmt)
    http://www.apache.org/licenses/LICENSE-2.0
controllers/awscluster_controller_test.go:8: File is not `gofmt`-ed with `-s` (gofmt)
    http://www.apache.org/licenses/LICENSE-2.0
controllers/helpers_test.go:8: File is not `gofmt`-ed with `-s` (gofmt)
    http://www.apache.org/licenses/LICENSE-2.0
pkg/cloud/services/s3/s3_test.go:22:2: SA1019: "io/ioutil" has been deprecated since Go 1.16: As of Go 1.16, the same functionality is now provided by package io or package os, and those implementations should be preferred in new code. See the specific function documentation for details. (staticcheck)
	"io/ioutil"
	^
level=info msg="File cache stats: 553 entries of total size 7.1MiB"
level=info msg="Memory: 1358 samples, avg is 694.6MB, max is 3263.8MB"
level=info msg="Execution took 2m34.066697983s"
make: *** [Makefile:280: lint] Error 1
```

dlipovetsky commented:
So, the prow job uses the `gcr.io/k8s-staging-test-infra/kubekins-e2e:v20220726-61b7bafca6-master` container image, and it has go 1.19rc2 installed? Using an rc seems very odd, but it looks like the CI [only cares about the minimum go version](https://github.com/kubernetes-sigs/cluster-api-provider-aws//blob/8e8f07ad4add0ee0e9dde07b5b2c251798e26925/scripts/ci-test.sh#L22-L23).

I can reproduce a `make lint` failure in this container image, but it's actually different from the one reported in the prow job:
```
hack/tools/bin/golangci-lint run -v --fast=false
INFO [config_reader] Config search paths: [./ /workspace/cluster-api-provider-aws /workspace / /root] 
INFO [config_reader] Used config file .golangci.yml 
INFO [lintersdb] Active 49 linters: [asasalint asciicheck bodyclose containedctx deadcode decorder depguard dogsled errcheck errchkjson execinquery exportloopref gci goconst gocritic gocyclo godot gofmt goheader goimports goprintffuncname gosec gosimple govet grouper ifshort importas ineffassign maintidx misspell nakedret nilerr noctx nolintlint nosprintfhostport prealloc predeclared revive rowserrcheck staticcheck structcheck stylecheck thelper typecheck unconvert unparam unused varcheck whitespace] 
INFO [loader] Using build tags: [tools e2e]       
INFO [loader] Go packages loading at mode 575 (exports_file|files|types_sizes|compiled_files|deps|imports|name) took 1.11034245s 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 55.042403ms 
INFO [linters context/goanalysis] analyzers took 0s with no stages 
INFO [runner] Issues before processing: 2040, after processing: 7 
INFO [runner] Processors filtering stat (out/in): skip_files: 910/2040, skip_dirs: 910/910, exclude: 412/905, nolint: 14/175, uniq_by_line: 7/14, source_code: 7/7, path_prefixer: 7/7, path_prettifier: 2040/2040, identifier_marker: 905/905, diff: 7/7, path_shortener: 7/7, sort_results: 7/7, cgo: 2040/2040, autogenerated_exclude: 905/910, exclude-rules: 175/412, max_same_issues: 7/7, max_from_linter: 7/7, filename_unadjuster: 2040/2040, max_per_file_from_linter: 7/7, severity-rules: 7/7 
INFO [runner] processing took 44.361829ms with stages: nolint: 13.011624ms, identifier_marker: 9.787978ms, autogenerated_exclude: 6.435139ms, path_prettifier: 4.074242ms, exclude-rules: 4.010892ms, exclude: 2.873986ms, skip_files: 2.342794ms, skip_dirs: 1.480255ms, cgo: 150.524µs, source_code: 94.678µs, filename_unadjuster: 90.3µs, max_per_file_from_linter: 3.426µs, uniq_by_line: 3.246µs, path_shortener: 1.393µs, max_same_issues: 541ns, severity-rules: 181ns, diff: 180ns, sort_results: 170ns, max_from_linter: 140ns, path_prefixer: 140ns 
INFO [runner] linters took 259.135747ms with stages: goanalysis_metalinter: 214.693374ms 
controlplane/eks/controllers/helpers_test.go:8: File is not `gofmt`-ed with `-s` (gofmt)
    http://www.apache.org/licenses/LICENSE-2.0
api/v1beta1/defaults.go:8: File is not `gofmt`-ed with `-s` (gofmt)
    http://www.apache.org/licenses/LICENSE-2.0
pkg/cloud/services/s3/mock_stsiface/doc.go:17: File is not `gofmt`-ed with `-s` (gofmt)
// Run go generate to regenerate this mock.
controllers/awscluster_controller_test.go:8: File is not `gofmt`-ed with `-s` (gofmt)
    http://www.apache.org/licenses/LICENSE-2.0
controllers/helpers_test.go:8: File is not `gofmt`-ed with `-s` (gofmt)
    http://www.apache.org/licenses/LICENSE-2.0
pkg/cloud/services/s3/mock_s3iface/doc.go:17: File is not `gofmt`-ed with `-s` (gofmt)
// Run go generate to regenerate this mock.
pkg/internal/rate/rate.go:213: File is not `gofmt`-ed with `-s` (gofmt)
//   r := lim.ReserveN(time.Now(), 1)
//   if !r.OK() {
//     // Not allowed to act! Did you remember to set lim.burst to be > 0 ?
//     return
//   }
//   time.Sleep(r.Delay())
//   Act()
INFO File cache stats: 7 entries of total size 54.2KiB 
INFO Memory: 16 samples, avg is 88.8MB, max is 149.4MB 
INFO Execution took 1.430546104s 
```

dlipovetsky commented:
/test pull-cluster-api-provider-aws-test

Skarlso commented:
Maybe we shouldn't be using `master` as the test image but pin it to a specific version? Using an RC version is super strange indeed.

richardcase commented:
Lets pin the jobs to a specific version and not master. I will make a change to test-infra now

richardcase commented:
This test failure should be fixed by: https://github.com/kubernetes/test-infra/pull/26929

richardcase commented:
/retest
